Pokergameapp final

okstring opened the pull request:
- #### 스텝별 작업 목록

- [x] 딜러와 참여자가 가진 카드를 요구사항에 맞춰서 계산해서 승자를 선택한다.
- [ ] 선택한 승자에 대한 표시를 화면에 출력한다.
  - 요구사항에서 `UIImageView` 를 추가를 원했지만 `UIImageView` 를 오토 레이아웃을 설정하는데에 원하는대로 나오지 않게 됐습니다.
  - 해당 부분 계속해서 진전이 없고 시간을 너무 많이 잡아먹어 지체될 것 같아 다음 미션 풀고난 후 개인적으로 해결해보려 합니다...........

- #### 학습 키워드

  - `CarIterable` : 잘못 사용하면 무한 재귀 호출이 되는 것을 알았습니다
  - `closure` : 지적해주신 부분들 closure를 사용해서 이전보다는 사용이 수월해졌습니다.
  - `constraint` : programmatically하게 작성하는 것에 대해 감을 잡았다 싶으면서도 제 마음대로 작동이 안되서 너어무 답답했습니다. 하나하나 뜯어보면서 연구가 필요해보입니다.

- #### 고민과 해결

  - `CaseIterable` 의 사용이 생소해서 이것저것 해보고 찾아봤습니다.
    - `lhs`, `rhs` 의 타입을 바꿔도 되는 건지: 다른 타입을 비교한다는 것 자체가 안정성 측면에서 말이 안되는 거였다
    - 왜 계속해서 호출이 되는지: 검색보다는 디버깅툴에서 해당 부분 살펴보고 수정했다. 유의가 필요해보였다.

- #### 부연 설명

  - **view 구조 설명**

    ![description](https://user-images.githubusercontent.com/62657991/108839547-8d7cef80-7618-11eb-9888-843377d626e8.png)

  

  - **승자 규칙 구현**

    - `Score` 라는 enum 타입을 추가했습니다. 안에는 `fourCard`, `straignt` 등 score 분류와 `Rank` 매개변수 값을 가질 수 있도록 했습니다.

    - `Dealer`가 score를 계산합니다. 같은 rank끼리 카운트해서 dictionary 형태로 담아둔 후 각각의 score 로직에 맞게 구현합니다.

    - **로직 내 변수 설명**(주석으로도 남겼습니다)

      - `maxCardScore`: eachCardScore 중 가장 높은 score를 Dictionary 형태로 가집니다

      - ` straightScore`: straignt 판정합니다. 아니라면 .none을 반환합니다

      - `correctTwoPairs` : TwoPair 판정을 위해 같은 카드가 두개인 것을 일단 골라냅니다.

okstring committed:
Card Image Name 변환 메소드 viewComtroller extension에 구현

okstring committed:
shareCards를 Dealer가 구현, PokerGame 매개변수 안받는 init 구현

okstring committed:
내부 배열에 접근하지 않도록 players.configEachPlayer 추가

okstring committed:
NumberOfParticipant.iterateRawValue closure 함수 사용

okstring committed:
SegmentedControl item 하드코딩하지 않도록 extension 수정

okstring committed:
NumberOfParticipant,init() 활용

okstring committed:
카드 게임 알고리즘 구현

okstring committed:
점수 높은 플레이어에게 메달 표시되도록 구현

okstring committed:
Clean code, README.md 작성

godrm commented on PokerGameApp/PokerGameApp/Model/Card.swift:
```diff
@@ -43,8 +44,9 @@ struct Card: CustomStringConvertible {
-        return "\(shape.imageName)\(rank)"
+    func calcurareScore(block: (Card.Rank) -> ()) {
+        let rank = self.rank
+        block(rank)
```
이 부분도 좋은 시도입니다. 다만 Rank 값까지 클로저로 처리할 필요는 없어 보입니다.
Card는 어차피 값만 갖고 있어서, Rank를 주고 판단하는 메소드 보다는 값을 비교하는 메소드를 잘 만드는 게 더 중요합니다.

okstring replied:
다섯번째 리뷰와 함께 수정했습니다 :)

godrm commented on PokerGameApp/PokerGameApp/Model/Dealer.swift:
```diff
@@ -2,13 +2,23 @@ import Foundation
     private var cards = [Card]()
+    private var cardDeck: CardDeck
+    var result: Score {
+        return Dealer.calcurateCardScore(cards: cards)
```
이런 메소드도 읽기 편하게 작성하는 방법을 고민해보세요.
`calcurateScore(of cards:)` 형태가 더 읽기 좋습니다. 단어가 반복되는 것도 주의하세요.

okstring replied:
더 좋고 깔끔합니다! 읽히기 쉬운 네이밍을 좀 더 고민해보겠습니다.

godrm commented on PokerGameApp/PokerGameApp/Model/Dealer.swift:
```diff
@@ -2,13 +2,23 @@ import Foundation
+    }
+    
+    init() {
+        self.cardDeck = CardDeck()
```
카드 덱이 필요하고 의존하면 init()할 때 넘겨줄 수 있는지 고민해보세요.

okstring replied:
Dealer init시 CardDeck을 받도록 구현했습니다

godrm commented on PokerGameApp/PokerGameApp/Model/Dealer.swift:
```diff
@@ -17,4 +27,82 @@ class Dealer: Playable {
+        }
+    }
+    
+    static func calcurateCardScore(cards: [Card]) -> Score {
```
아.. 이게 static 타입 메소드네요. 이 타입 (혹은 인스턴스와) 크게 관련이 없다는 의미겠죠?
그렇다면 이 함수는 어디에 있는게 좋을까요? 
카드 배열을 가져가서 점수를 만드는 함수니까 카드 배열을 소유하고 보관하는 객체에 있으면 어떨까요?

okstring replied:
static을 쓰면서도 이게 맞는건지 망설이게 됩니다. 말씀하신대로 관련이 없고 Dealer가 계산하는 것이라 접근, 메소드 접근하기 용이해서 썼지만 자주 쓰면 객체지향적이지 않게 되고 해당 객체 라이프타임이 계속간다는 것을 알았습니다. static 쓰는것을 지양하는것이 좋을지 궁금합니다.

godrm commented on PokerGameApp/PokerGameApp/Model/Dealer.swift:
```diff
@@ -17,4 +27,82 @@ class Dealer: Playable {
+    static func calcurateCardScore(cards: [Card]) -> Score {
+        var eachCardScore = Dictionary<Card.Rank, Int>()
+        for card in cards {
+            card.calcurareScore { (rank) in
```
혹시 calcurare 라는 단어가 있나요? 어떤 의도인지 모르겠습니다.
카드 하나를 점수화하는 건가요? 이 정도면 그냥 card.score() 해도 되지 않을까요. 
Card는 카드게임이나 플레이어처럼 비즈니스 로직을 갖고 있는 것도 아니고 여기저기 전달되는 타입이라서 
private var 가 아니라 private (set) 또는 let 도 괜찮습니다. Card 내부 값이 바뀌지는 않으니까요. 
그러면 이렇게까지 하지 않아도 되는거죠.

okstring replied:
오타가 났던 것 같습니다 더욱 꼼꼼하게 보겠습니다!

승패 알고리즘을 구현할 때 가지고 있는 card의 rank를 카운트해서 fourCard부터 순차적으로 해당 스코어 맞는지 확인하는 방식으로 작성했습니다. 
이때 rank의 카운트가 필요해서 get 메소드가 아닌 클로저 형식으로 rank를 담아서 보내는 방식으로 구현했지만 get메소드와 모양이 비슷해보여서 이래도 되는것인지 궁금했었습니다. 이런 고민은 private(set)으로 바꿀수 있어 마음이 편안해진 것 같습니다😄

+ Comparable은 어디서 쓰였나면 카운트를 다 하고 그 중 큰 값을 구할 때(`eachCardScore.max` 안에서) 썼습니다.
+ 네이밍에서도 혼동이 있을 것 같아 `countScoreOfRankValue` 바꿨습니다.

godrm commented on PokerGameApp/PokerGameApp/Model/Dealer.swift:
```diff
@@ -17,4 +27,82 @@ class Dealer: Playable {
+        default:
+            return .none
+        }
+    }
```
점수를 계산하는 로직이 꼭 Dealer에 있어야 하는지 한 번 고민해보세요.

godrm commented on PokerGameApp/PokerGameApp/Model/Player.swift:
```diff
@@ -3,6 +3,9 @@ import Foundation
     let name: String
     private var cards = [Card]()
+    var result: Score {
+        return Dealer.calcurateCardScore(cards: cards)
```
딜러가 점수 계산하는 로직을 갖고 있으니 Player가 Dealer 메소드를 호출하는 상황이 되서 이상하네요.

godrm commented on PokerGameApp/PokerGameApp/Model/Players.swift:
```diff
@@ -2,22 +2,19 @@ import Foundation
             players.append(player)
         }
-        let dealer = Dealer()
         players.append(dealer)
```
잘 생각해보세요. Player는 Playable 이라는 프로토콜로 추상화해서 배열로 갖고 있는데, Dealer는 구체 타입을 그냥 참조해야 할까요?
여기서 Dealer 타입 대신에 Playable 프로토콜 타입을 지정해도 될까요?

okstring replied:
`Dealer는 구체 타입을 그냥 참조해야 할까요?` 의 의미가 그냥 참조한다는것이 Dealer로 받기보다는 playable로 받아서 배열과 타입을 통일시키는 것이 좋을 것이라는 말씀으로 이해가 됐습니다. 딜러를 넣어야 한다는 생각에 깊게 생각하지 못했던 것 같습니다. 매개변수로 받을 때 Playable로 수정했습니다!

godrm commented on PokerGameApp/PokerGameApp/Model/Players.swift:
```diff
@@ -26,4 +23,25 @@ class Players {
+        }
+    }
+    
+    public func winnerPlayerSeatIndex() -> Int {
```
이 메소드도 좋은 접근이지만 Index를 그대로 넘겨주는 것보다는 어떤 구체적인 타입을 넘겨줄 수는 있지 않을까 생각해보세요.

okstring replied:
좌석 index가 아닌 playable 타입을 반환하는 것으로 반환했습니다. 약간의 요행으로 index를 반환했지만.. 💦 이렇게 반환하는 것이 자연스러워 보입니다.

godrm commented on PokerGameApp/PokerGameApp/Model/Score.swift:
```diff
@@ -0,0 +1,55 @@
+        case .straight: return "straight"
+        case .fourCard: return "fourCard"
+        default: return "none"
+        }
```
객체 내부에서만 쓰는 건가요? 외부에서도 쓰는 건가요? 외부에서도 쓰는 거라면 rawValue라는 computed property 보다는 차라리 description을 쓰세요.

okstring replied:
description으로 수정 완료했습니다. 타입을 받는 enum이라 rawValue를 생성하고야 말겠다는 생각이 강했던 것 같습니다 😓

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -33,38 +35,76 @@ class ViewController: UIViewController {
             let playerNameLabel = setPlayerNameLabel(name: player.name)
             let imageStackView = setImageStackView(stud: stud, player: player)
-            addPlayersStackViewIntoPokerPlate(imageStackView: imageStackView, nameLabel: playerNameLabel, numberOfPlayer: numberOfPlayer)
+            addPlayersResultStackViewIntoPokerPlate(imageStackView: imageStackView, nameLabel: playerNameLabel, numberOfPlayer: numberOfPlayer)
```
와.. 함수 이름 기네요. StackView는 뒤에 매개변수도 있으니까 생략하면 어떨까요.
스위프트는 함수 이름을 매개변수 포함이라고 생각하세요.

okstring replied:
매개변수도 네이밍에 포함시켜 지어보도록 하겠습니다!

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -33,38 +35,76 @@ class ViewController: UIViewController {
+            stackView.addArrangedSubview(winnerSignLabel)
+            stackView.heightAnchor.constraint(equalToConstant: 90).isActive = true
+            return stackView
+        }()
```
함수 내부에서 쓸 때는 클로저 형태로 안쓰고 let stackView 부터 그냥 써도 됩니다.

okstring replied:
이게 더 자연스러워보입니다! 수정 완료했습니다.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -218,3 +258,20 @@ class ViewController: UIViewController {
+    }
+    
+    var cardsSegmentedControlItemsName: Array<String> {
+        CardStud.allCases.map{ "\($0.rawValue) Cards" }
```
computed property는 매번 호출할 때마다 이 값을 새로 생성합니다. 
자주 사용하는 경우는 조심해서 사용하세요.

okstring replied:
네! Computed property 사용할 때 주의해서 사용하겠습니다.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -218,3 +258,20 @@ class ViewController: UIViewController {
+            .replacingOccurrences(of: "♣️", with: "c")
+            .replacingOccurrences(of: "♦️", with: "d")
+            .replacingOccurrences(of: "❤️", with: "h")
+            .replacingOccurrences(of: "♠️", with: "s")
```
이번 미션은 괜찮지만 다른 경우는 description을 로직으로 활용하지는 마세요.

okstring replied:
앞으로는 description이 하는 역할에 좀 더 중점을 두도록 하겠습니다 :)

godrm reviewed (commented):
수고하셨습니다. 마무리 단계까지 다 하셨네요.

okstring commented:
주신 리뷰 읽어보고 부끄럽게 놓치고 있는 부분이 많다고 느꼈습니다. 내일 좀 더 꼼꼼하게 보고 수정해본 후 다시 push하도록 하겠습니다 늦은 시간에 감사드립니다 :)

okstring committed:
Card.countScoreOfRankValue 로 이름 변경, Dealer.init()시 CardDeck을 받도록 구현

okstring committed:
card 속성 private(set)으로 변경, Playable.swift 생성

 - 플레이어 등록 시 dealer 매개변수를 Playable 수정
 - Playable안에서 카드 점수 구현할 수 있는 extension 구현,

okstring committed:
winnerPlayer 메소드 반환타입 변환, Score description 구현

 - 타입 변환에 따른 viewController 다르게 구현

okstring committed:
함수 내 StackView 등 클로저 형식 선언 아닌 직접 속성 설정으로 변경

okstring commented:
수정 후 push 완료했습니다! 많은 부분 지적해주시고 수정이 필요한 걸 보고 당분간은 이게 옳은 방향이라는걸 제 자신에게 인식시키는 시간이 필요해보일 듯 합니다. 감사합니다 :)

godrm commented on PokerGameApp/PokerGameApp/Model/Playable.swift:
```diff
@@ -0,0 +1,79 @@
+        var eachCardScore = Dictionary<Card.Rank, Int>()
+        for card in cards {
+            eachCardScore[card.rank, default: 0] += 1
+        }
```
Playable 프로토콜에 구현한다는 것은 이 프로토콜을 채택하는 타입은 
모두 해당 메소드를 처리한다는 걸 의미하구요. 추상화된 타입에 구현을 넣어놓는 셈이죠.
그래서 정말 공통적으로 다 쓰는 메소드만 여기에 구현하면 됩니다. 지금도 나쁘지 않네요
별도 타입으로 선언할꺼냐. 공통된 기능을 모아서 추상화 또는 일반화할꺼냐 고민은 끝이 없습니다.

godrm commented on PokerGameApp/PokerGameApp/Model/Players.swift:
```diff
@@ -30,7 +30,7 @@ class Players {
         }
     }
     
-    public func winnerPlayerSeatIndex() -> Int {
```
이거 좋네요 :)
